### PR TITLE
Move certificates functions into a new file

### PR DESCRIPTION
**Pull request by draychev**
This is a piece of the larger refactor here https://github.com/Azure/application-gateway-kubernetes-ingress/pull/156

In this PR we carve out some of the existing code into a new file - `certificates.go`. This provides:

  1. function getSslCertificates which returns `*[]ApplicationGatewaySslCertificate` list
  2. function getCertificateV1, which given a hostname returns a base64 cert/secretID.

The code here won't be used until the rest of  https://github.com/Azure/application-gateway-kubernetes-ingress/pull/156 is merged and as such is dead code (or duplication of the existing functionality)

The goal of this overall refactor is to remove functions with side-effects and the requirement to execute these in specific order to build up `appGwConfigBuilder` data structures.

### CODE REVIEW

File: pkg/appgw/certificates.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+		}
+
+		tlsSecret := secretIdentifier{
+			Name:      tls.SecretName,
+			Namespace: ingress.Namespace,
+		}
+
+		// add hostname-tlsSecret mapping to a per-ingress map
+		cert := builder.k8sContext.CertificateSecretStore.GetPfxCertificate(tlsSecret.secretKey())
+		if cert == nil {
```
Comment (asridharan) [inline pkg/appgw/certificates.go]: We don't need this right? We could just replace this by :
```
if cert != nil {
secretIDCertificateMap[tlsSecret] = to.StringPtr(base64.StdEncoding.EncodeToString(cert))
}

**Comment by akshaysngupta** [comment]
certificate.go serves two purposes:
1) provide function `getSslCertificates` which returns `*[]ApplicationGatewaySslCertificate` list
2) provides function `getCertificateV1`, that given a hostname returns base64 cert/secretID.

Can you add a summary for these two ?

File: pkg/appgw/certificates.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+	cert, exists := builder.getSecretToCertificateMap(ingress)[secID]
+	if !exists {
+		// secret referred does not correspond to a certificate
+		return nil, nil
+	}
+	return cert, &secID
+}
+
+func (builder *appGwConfigBuilder) makeHostToSecretMap(ingress *v1beta1.Ingress) map[string]secretIdentifier {
```
Comment (asridharan) [inline pkg/appgw/certificates.go]: Should we follow the `new` idiom. So s/makeHostToSecretMap/newHostToSecretMap ?

File: pkg/appgw/certificates.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+				hostToSecretMap[""] = tlsSecret
+			} else {
+				hostToSecretMap[hostname] = tlsSecret
+			}
+		}
+	}
+	return hostToSecretMap
+}
+
+func makeCert(secretID secretIdentifier, cert *string) network.ApplicationGatewaySslCertificate {
```
Comment (asridharan) [inline pkg/appgw/certificates.go]: s/makeCert/newCert ?

File: pkg/appgw/certificates_test.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	"k8s.io/client-go/tools/cache"
+)
+
+func TestMakeHostToSecretMap(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Test setting up SSL redirect annotations")
+}
+
+var _ = Describe("makeHostToSecretMap", func() {
+	Context("looking at TLS specs", func() {
```
Comment (asridharan) [inline pkg/appgw/certificates_test.go]: This context seems odd? Maybe :
` Testing caching of secrets for ingress with TLS spec`?

File: pkg/appgw/certificates_test.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	. "github.com/onsi/gomega"
+	"k8s.io/client-go/tools/cache"
+)
+
+func TestMakeHostToSecretMap(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Test setting up SSL redirect annotations")
+}
+
+var _ = Describe("makeHostToSecretMap", func() {
```
Comment (asridharan) [inline pkg/appgw/certificates_test.go]: s/makeHostToSecretMap/Testing `makeHostToSecretMap`

File: pkg/appgw/certificates_test.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+	"github.com/Azure/application-gateway-kubernetes-ingress/pkg/annotations"
+	"github.com/Azure/application-gateway-kubernetes-ingress/pkg/k8scontext"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"k8s.io/client-go/tools/cache"
+)
+
+func TestMakeHostToSecretMap(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Test setting up SSL redirect annotations")
```
Comment (asridharan) [inline pkg/appgw/certificates_test.go]: The RunSpecs description is probably invalid?

File: pkg/appgw/certificates_test.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+				testFixturesNamespace,
+				testFixturesNameOfSecret,
+			},
+		}
+		It("should succeed", func() {
+			Expect(actual).To(Equal(expected))
+		})
+	})
+
+	Context("running", func() {
```
Comment (asridharan) [inline pkg/appgw/certificates_test.go]: Can we be more descriptive with the `Context` description?

File: pkg/appgw/certificates_test.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+			Expect(secrets[testFixturesHost]).To(Equal(expected))
+		})
+	})
+
+	Context("extract host to secret map", func() {
+		certs := getCertsTestFixture()
+		cb := makeConfigBuilderTestFixture(&certs)
+
+		ingress := makeIngressFixture()
+		ingress.Annotations[annotations.SslRedirectKey] = "true"
```
Comment (asridharan) [inline pkg/appgw/certificates_test.go]: Why do we need this annotation?

File: pkg/appgw/certificates_test.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+		It("has the correct secrets", func() {
+			expected := secretIdentifier{
+				Namespace: testFixturesNamespace,
+				Name:      testFixturesNameOfSecret,
+			}
+			Expect(secrets[testFixturesHost]).To(Equal(expected))
+		})
+	})
+
+	Context("extract host to secret map", func() {
```
Comment (asridharan) [inline pkg/appgw/certificates_test.go]: Maybe need a bit more description here? Trying to understand the difference between this test and the test above "looking at TLS spec"?

**Pull request merged**